tests/test_utils.py

```diff
@@ -35,3 +62,50 @@ def test_memory_human_readable_conversion():
     assert bytes_to_human_readable(1024) == (1.0, "KB")
     assert bytes_to_human_readable(1024**2) == (1.0, "MB")
     assert bytes_to_human_readable(4.3 * 1024**4) == (4.3, "TB")
+
+
+@pytest.mark.parametrize(
+    ("target, hook, error"),
+    [
+        param(
+            f_annotated,
+            f_annotated_one_default,
+            None,
+            id="hook_with_defaults",
+        ),
+        param(
+            f_annotated_one_default,
+            f_annotated,
+            None,
+            id="target_with_defaults",
+        ),
+        param(
+            f_annotated,
+            f_plain,
+            None,
+            id="hook_without_annotations",
+        ),
+        param(
+            f_annotated,
+            f_reversed_annotated,
+            TypeError,
+            id="different_order",
+        ),
+        param(
+            f_annotated,
+            f2_plain,
+            TypeError,
+            id="different_names",
+        ),
+        param(
+            f_annotated,
+            f_reduced_plain,
+            TypeError,
+            id="hook_missing_arguments",
+        ),
+    ],
+)
+def test_register_hook(target, hook, error):
+    """Passing in-/consistent signatures to `register_hook` raises an/no error."""
+    with pytest.raises(error) if error is not None else nullcontext():
+        register_hooks(target, [hook])
```

Comment:
Could you please mention the missing test cases (I'm just curious abut them).

Reply:
Sure (you'll also find the details in the commit history):

#### Missing/Failing Case 1
Check for unequal number of parameters (the `zip` would only iterate until the shorter iterable is exhausted)

#### Missing/Failing Case 2
When the parameter names differed but the parameter was not annotated in the hook. That slipped through because the conditions where not cleanly separated in your validation logic

#### Unnecessary Cases
Your `target_has_different_signature_order` and `target_has_different_names` were not needed because they test exactly the same thing as the corresponding cases starting with `hook_has_*`

Reply:
Okay I see, I did not expect the first one. Thank you!